### PR TITLE
Updated contracts

### PR DESCRIPTION
**Pull request by SoIidarity**
This pull request marks a rather large update to the Radi Cards contract code base. Over and above the initial functionality a whole load of new shiny things have been added. The original contract interfaces were left intact as much as possible while adding new functionality to minimize the changes required within the store on the front end. Inspiration was taken from the cryptoxmas contracts for the claimable links but I did take some small design decisions in the implementation that differed from theirs.
 
I had a lot of features to implement in a short amount of time and so we are still primarily using one contract with a lot of main unit tests. Ideally, future versions should involve an upgradability pattern with separate storage and logic layers and a more modular contract and testing architecture. Either way, I feel this is fine for now.
 
The main features that were implemented in this PR are:

- Ability to send ETH to recipient as well as the charity.
     - User can choose any ratio between the recipient and charity
- Ability to send DAI to recipient and charity
     - User can choose any ratio as in this config as well
     - The address of the dai erc20 contract must be set after contract deployment
- Introduction of limited supply cards
     - We can now set the maximum number that each card can be minted.
- Introduction of minimum price per token
     - This clearly works well with the limited supply card
     -  Basic logic is the sum of donation and gift must be more than the defined  card min
     - The price of each token is pulled from the makerdao oracle!
          - This needs to be set after the contract is deployed
- Introduction of claimable link logic!
     - Basically, links are generated with the cryptoxmas volca logic.
     - Can still sent do an address, if you know it.
     - Card and gift sit in escro within the main contract until claimed
     - Ephemeral account is setup and is used to claim the nft and the gift
     - If the gift is cancelled the gift and nft is sent back to the gifter but the donation is not recoverable

 
I suspect someone will say I should join the gift and giftInDai functions. The logic in keeping these separate is they actually do quite very different things interms of their logic and how they handle the funds sent into them. Their parameters are also different in the sense that the donation and gift amount for the one is in eth and the other dai. This makes the interfaces confusing. If we joined them we would need logic in the front end to generate different kinds of contract calls in any case to handle both the eth and dai donations.
 
Interms of commenting on the contracts I’ve not followed the standard patters due to restrictions in time and have only left comments on functions that really needed them.
 
That basically sums it up. Feedback is most welcome. I know the tests are a bit complex and over the place but I really was trying to fit everything into the original structure to safe time where possible. Ideally, we should bump our zeplin solidity version and the solidity version its self but this is fine for now. No real downside with using this version.

### CODE REVIEW

**Comment by SoIidarity** [comment]
I just ran the tests and saw that some of them were failing....it appears to be the same bug I was experiencing yesterday and fought with for 3 hours. it was resolved by adding `ERC721Holder` to the contract implementation but it seems to have reappeared. I will look into this at some point.

File: contracts/RadiCards.sol [hunk 1]
```diff
@@ -3,234 +3,504 @@ pragma solidity ^0.4.24;
...
-  returns (bool) {
-    require(bytes(_tokenURI).length != 0, "Invalid token URI");
-
-    cards[_cardIndex] = CardDesign(
-      _tokenURI,
-      _active
+    event LogClaim(
+            address indexed ephemeralAddress,
+            address indexed sender,
+            uint tokenId,
```
Comment (jamesmorgan) [inline contracts/RadiCards.sol]: Maybe add `indexed` to the `tokenId` as the final and 3rd index on this event

File: contracts/RadiCards.sol [hunk 2]
```diff
@@ -3,234 +3,504 @@ pragma solidity ^0.4.24;
...
-
-  function setTokenBaseURI(string _newBaseURI) external onlyIfWhitelisted(msg.sender) {
-    require(bytes(_newBaseURI).length != 0, "Base URI invalid");
 
-    tokenBaseURI = _newBaseURI;
-  }
+    event LogCancel(
+        address indexed ephemeralAddress,
+        address indexed sender,
+        uint tokenId
```
Comment (jamesmorgan) [inline contracts/RadiCards.sol]: Same on this

File: contracts/RadiCards.sol [hunk 3]
```diff
@@ -3,234 +3,504 @@ pragma solidity ^0.4.24;
...
+            to.transfer(EPHEMERAL_ADDRESS_FEE);
+        }
+
+        else {
+            require(_donationAmount + _giftAmount == msg.value,"Can only request to donate and gift the amount of ether sent");
+            _giftStatus = Statuses.Claimed;
+            _sentToAddress = to;
+        }
+
+        if (cards[_cardIndex].maxQnty > 0){ //the max quantity is set to zero to indicate no limit. Only need to check that can mint if limited
```
Comment (jamesmorgan) [inline contracts/RadiCards.sol]: Should this check happen earlier and should it not be a `modifier` so we can save some gas?

File: contracts/RadiCards.sol [hunk 4]
```diff
@@ -3,234 +3,504 @@ pragma solidity ^0.4.24;
...
+            require(cards[_cardIndex].minted < cards[_cardIndex].maxQnty, "Can't exceed maximum quantity of card type");
+        }
+
+        if(cards[_cardIndex].minPrice > 0){ //if the card has a minimum price check that enough has been sent
```
Comment (jamesmorgan) [inline contracts/RadiCards.sol]: Also this may be better as a `modifier` as well

File: contracts/RadiCards.sol [hunk 5]
```diff
@@ -3,234 +3,504 @@ pragma solidity ^0.4.24;
...
+
+    function giftInDai(address to, uint256 _benefactorIndex, uint256 _cardIndex, string _message, uint256 _donationAmount, uint256 _giftAmount, bool _claimableLink) public payable returns (bool) {
+        require(to != address(0), "Must be a valid address");
+        require(benefactors[_benefactorIndex].ethAddress != address(0), "Must specify existing benefactor");
+        require(bytes(cards[_cardIndex].tokenURI).length != 0, "Must specify existing card");
+        require(cards[_cardIndex].active, "Must be an active card");
+        require((_donationAmount + _giftAmount)<= daiContract.allowance(msg.sender, this), "Must have provided high enough alowance to Radicard contract");
+        require((_donationAmount + _giftAmount)<= daiContract.balanceOf(msg.sender), "Must have enough token balance of dai to pay for donation and gift amount");
+
+        if (cards[_cardIndex].maxQnty > 0){ //the max quantity is set to zero to indicate no limit. Only need to check that can mint if limited
```
Comment (jamesmorgan) [inline contracts/RadiCards.sol]: Replace with modifier as mentioned above

File: contracts/RadiCards.sol [hunk 6]
```diff
@@ -3,234 +3,504 @@ pragma solidity ^0.4.24;
...
+            require(cards[_cardIndex].minted < cards[_cardIndex].maxQnty, "Can't exceed maximum quantity of card type");
+        }
+
+        if(cards[_cardIndex].minPrice > 0){ //if the card has a minimum price check that enough has been sent
```
Comment (jamesmorgan) [inline contracts/RadiCards.sol]: Replace with modifier as mentioned above

File: contracts/RadiCards.sol [hunk 7]
```diff
@@ -3,234 +3,504 @@ pragma solidity ^0.4.24;
...
+            require((_donationAmount + _giftAmount) >= cards[_cardIndex].minPrice, "The total dai sent with the transaction is less than the min price of the token");
+        }
+
+        //parameters that change based on the if the card is setup as with a claimable link
+        Statuses _giftStatus;
+        address _sentToAddress;
+
+        if(_claimableLink){
```
Comment (jamesmorgan) [inline contracts/RadiCards.sol]: Not 100% sure of my understanding of the claimable link stuff, but does this mean that it sends them  `DAI` of `EPHEMERAL_ADDRESS_FEE` amount and not ETH - is this correct and as designed?

File: contracts/RadiCards.sol [hunk 8]
```diff
@@ -3,234 +3,504 @@ pragma solidity ^0.4.24;
...
+    uint256 public totalGiftedInAtto; //SI prefix for 1x10-18 dai is Atto.
+    uint256 public totalDonatedInAtto;
+
+    event CardGifted(
+        address indexed _to,
+        uint256 indexed _benefactorIndex,
+        uint256 indexed _cardIndex,
+        address _from,
+        uint256 _tokenId,
+        Statuses status
```
Comment (jamesmorgan) [inline contracts/RadiCards.sol]: Is there a need to understand if the card was gift with `DAI` or `ETH`, if so then maybe add a flag to the event.

File: contracts/RadiCards.sol [hunk 9]
```diff
@@ -3,234 +3,504 @@ pragma solidity ^0.4.24;
...
+            else{
+                msg.sender.transfer(card.giftAmount);
+            }
+        }
+
+        // send nft to buyer
+        super.transferFrom(this, msg.sender, tokenId);
+
+        // log cancel event
+        emit LogCancel(_ephemeralAddress, msg.sender, tokenId);
```
Comment (jamesmorgan) [inline contracts/RadiCards.sol]: Maybe make the event more descriptive, cancelling what?

File: contracts/RadiCards.sol [hunk 10]
```diff
@@ -3,234 +3,504 @@ pragma solidity ^0.4.24;
...
+            if(card.daiDonation){
+                require(daiContract.transfer(_receiver, card.giftAmount),"Sending to recipient after cancel gift failed");
+        }
+            else{
+                _receiver.transfer(card.giftAmount);
+            }
+        }
+
+        // log claim event
+        emit LogClaim(
```
Comment (jamesmorgan) [inline contracts/RadiCards.sol]: Again maybe tweak to be more descriptive.

File: contracts/RadiCards.sol [hunk 11]
```diff
@@ -3,234 +3,504 @@ pragma solidity ^0.4.24;
...
+    }
+
+  //returns the value of ether in atto  (1 usd of ether = 10^18 atto)
+    function getEthUsdValue(uint256 _ether) public view returns(uint256){
+        return ((_ether*getEtherPrice())/(1 ether));
+    }
+
+  //returns the minimum required in wei for a particular card given the card min price in dai
+    function getMinCardPriceInWei(uint256 _cardIndex) public view returns(uint256){
+        return ((cards[_cardIndex].minPrice*1 ether)/getEtherPrice());
```
Comment (jamesmorgan) [inline contracts/RadiCards.sol]: Looks good, alot of functionality has been added and from first look it seem great. I would make you are happy enough with the test coverage and some of the edge cases around claiming etc so we can be confident there isn't a sneaky little bug. Top banana 🔝 🍌

File: contracts/mocks/MedianizerMock.sol [hunk 12]
```diff
@@ -0,0 +1,383 @@
...
+    }
+
+    function div(uint256 x, uint256 y) constant internal returns (uint256 z) {
+        z = x / y;
+    }
+
+    function min(uint256 x, uint256 y) constant internal returns (uint256 z) {
+        return x <= y ? x : y;
+    }
+    function max(uint256 x, uint256 y) constant internal returns (uint256 z) {
```
Comment (jamesmorgan) [inline contracts/mocks/MedianizerMock.sol]: these all look like `safemath` to me, I wonder if we can switch them out. No worries if not

File: contracts/RadiCards.sol (hunk 1 above)
Comment (SoIidarity) [reply contracts/RadiCards.sol]: done

File: test/truffle/ERC721MintBurn.behavior.js [hunk 13]
```diff
@@ -26,16 +32,17 @@ function shouldBehaveLikeMintAndBurnERC721 (
 
   const TOKEN_URI = '123abcHash';
 
+  const oneUSDInWei = etherToWei(1).dividedToIntegerBy(130);
```
[the remark this answers is not in the document]
Comment (SoIidarity) [reply test/truffle/ERC721MintBurn.behavior.js]: indeed. I explain this logic in the RadiCards.custom.test.js file

File: test/truffle/RadiCards.custom.test.js [hunk 14]
```diff
@@ -65,169 +86,1197 @@ contract('RadiCards ERC721 Custom', function (accounts) {
...
+          this.token.address
+        );
+        contractBalanceAfter.should.be.bignumber.equal(contractBalanceBefore.minus(oneUSDInAtto))
+
+        // we want to claim the token and eth to account2
+        let recipientBalanceAfter = await this.daiContract.balanceOf(
+          account2
+        );
+        recipientBalanceAfter.should.be.bignumber.equal(recipientBalanceBefore.plus(oneUSDInAtto))
+      });
```
[the remark this answers is not in the document]
Comment (SoIidarity) [reply test/truffle/RadiCards.custom.test.js]: will check the coverage now and fill in the lacking sections

File: test/test_truffle.sh [hunk 15]
```diff
@@ -62,5 +62,5 @@ if [ "$SOLIDITY_COVERAGE" = true ]; then
     cat coverage/lcov.info | node_modules/.bin/coveralls
   fi
 else
-  node_modules/.bin/truffle test --show-events "$@"
+  node_modules/.bin/truffle test test/truffle/RadiCards.custom.test.js --show-events "$@"
```
[the remark this answers is not in the document]
Comment (SoIidarity) [reply test/test_truffle.sh]: will do so. my bad.

File: contracts/RadiCards.sol (hunk 2 above)
Comment (SoIidarity) [reply contracts/RadiCards.sol]: done

File: contracts/RadiCards.sol (hunk 8 above)
Comment (SoIidarity) [reply contracts/RadiCards.sol]: I agree. I had this originally but the event was too big, with too many params and was failing. will look into optimizing it

File: contracts/RadiCards.sol (hunk 7 above)
Comment (SoIidarity) [reply contracts/RadiCards.sol]: nope. this should send the ephemeral address the `EPHEMERAL_ADDRESS_FEE ` in ETH as its calling the .transfer function on the address. this is verified in the unit tests

File: contracts/mocks/MedianizerMock.sol (hunk 12 above)
Comment (SoIidarity) [reply contracts/mocks/MedianizerMock.sol]: so I didnt write the `medinizerMock.sol`. this is the current implementation used by makerdao so you'll see when you run a `truffle compile` it throws a tone of warnings of deprecated functions and methods due to the old version of solidity that they used.

File: contracts/RadiCards.sol [hunk 16]
```diff
@@ -3,234 +3,504 @@ pragma solidity ^0.4.24;
 import "openzeppelin-solidity/contracts/math/SafeMath.sol";
 import "openzeppelin-solidity/contracts/access/Whitelist.sol";
 import "openzeppelin-solidity/contracts/token/ERC721/ERC721Token.sol";
+import "openzeppelin-solidity/contracts/token/ERC721/ERC721Holder.sol";
+import "openzeppelin-solidity/contracts/token/ERC20/StandardToken.sol";
 
 import "./Strings.sol";
+import "./Medianizer.sol";
```
Comment (andygray) [inline contracts/RadiCards.sol]: can this be pulled from MakerDAO rather than copying in?

File: contracts/RadiCards.sol [hunk 17]
```diff
@@ -3,234 +3,504 @@ pragma solidity ^0.4.24;
...
+* telergram, wechat ect. and is embedded within a URL to a claim page. When the user
+* opens the claim URL the private key is used to redeem the NFT and any associated gift (ETH/DAI).
+
 * @author blockrocket.tech (smart contracts)
 * @author cryptodecks.co
 * @author knownorigin.io
 * @author pheme.app
 * @author d1labs.com
 * @author mbdoesthings.com
+* @author chrismaree.io (smart contracts v2 update)
```
Comment (andygray) [inline contracts/RadiCards.sol]: 👍

File: contracts/RadiCards.sol [hunk 18]
```diff
@@ -3,234 +3,504 @@ pragma solidity ^0.4.24;
...
 
-    cards[_cardIndex].active = _active;
-  }
+    constructor () public ERC721Token("RadiCards", "RADI") {
+        addAddressToWhitelist(msg.sender);
+    }
+
+    function gift(address to, uint256 _benefactorIndex, uint256 _cardIndex, string _message, uint256 _donationAmount, uint256 _giftAmount, bool _claimableLink) payable public returns (bool) {
+        require(to != address(0), "Must be a valid address");
+        require(benefactors[_benefactorIndex].ethAddress != address(0), "Must specify existing benefactor");
```
Comment (andygray) [inline contracts/RadiCards.sol]: so we need to pick a benefactor even if we aren't sending anything to them? We could remove the `require` as the UI or whatever may just want to pass `address(0)` when not relevant?

Comment (SoIidarity) [reply contracts/RadiCards.sol]: I agree with this logic. will implement

File: contracts/RadiCards.sol (hunk 4 above)
Comment (andygray) [reply contracts/RadiCards.sol]: @SoIidarity I think it should be `<` and not `<=` right as you can send the exact min amount?

Comment (andygray) [reply contracts/RadiCards.sol]: makes more sense to my personally the other way `msg.value >= minPrice`

File: contracts/RadiCards.sol [hunk 19]
```diff
@@ -3,234 +3,504 @@ pragma solidity ^0.4.24;
...
+        tokenIdToRadiCardIndex[tokenIdPointer] = RadiCard({
+            gifter : msg.sender,
+            message : _message,
+            daiDonation: false,
+            giftAmount : _giftAmount,
+            donationAmount: _donationAmount,
+            status: _giftStatus,
+            cardIndex : _cardIndex,
+            benefactorIndex : _benefactorIndex
+        });
```
Comment (andygray) [inline contracts/RadiCards.sol]: combo of style with `:` - I would always do `blah:` rather than `blah :`

File: contracts/RadiCards.sol [hunk 20]
```diff
@@ -3,234 +3,504 @@ pragma solidity ^0.4.24;
...
+
+        // Card is minted to the _sentToAddress. This is either this radicards contract(if claimableLink==true)
```
Comment (andygray) [inline contracts/RadiCards.sol]: the comments here are not clarifying it for me. Maybe just say "this could be a ephermeral address"

File: contracts/RadiCards.sol [hunk 21]
```diff
@@ -3,234 +3,504 @@ pragma solidity ^0.4.24;
...
+            benefactors[_benefactorIndex].ethAddress.transfer(_donationAmount);
+        }
+        // transfer gift to recipient.
+        // note that we only do this if the link is not claimable as if it is the eth sits in escrow within this contract
+        if(_giftAmount > 0 && !_claimableLink){
+            _sentToAddress.transfer(_giftAmount);
+        }
+
+        // tally up the total eth gifted and donated
+        totalGiftedInWei = totalGiftedInWei.add(_giftAmount);
```
Comment (andygray) [inline contracts/RadiCards.sol]: you could add this inside the `if` statement to save gas and zero calls

File: contracts/RadiCards.sol [hunk 22]
```diff
@@ -3,234 +3,504 @@ pragma solidity ^0.4.24;
...
+        totalDonatedInWei = totalDonatedInWei.add(_donationAmount);
+
+        emit CardGifted(_sentToAddress, _benefactorIndex, _cardIndex, msg.sender, _tokenId, _giftStatus);
+        return true;
+    }
+
+    function giftInDai(address to, uint256 _benefactorIndex, uint256 _cardIndex, string _message, uint256 _donationAmount, uint256 _giftAmount, bool _claimableLink) public payable returns (bool) {
+        require(to != address(0), "Must be a valid address");
+        require(benefactors[_benefactorIndex].ethAddress != address(0), "Must specify existing benefactor");
```
Comment (andygray) [inline contracts/RadiCards.sol]: same comment as above

File: contracts/RadiCards.sol [hunk 23]
```diff
@@ -3,234 +3,504 @@ pragma solidity ^0.4.24;
...
+        uint256 indexed _cardIndex
+    );
 
-    cards[_cardIndex].active = _active;
-  }
+    constructor () public ERC721Token("RadiCards", "RADI") {
+        addAddressToWhitelist(msg.sender);
+    }
+
+    function gift(address to, uint256 _benefactorIndex, uint256 _cardIndex, string _message, uint256 _donationAmount, uint256 _giftAmount, bool _claimableLink) payable public returns (bool) {
```
Comment (andygray) [inline contracts/RadiCards.sol]: use `_to` in this method and below for consistency

File: contracts/RadiCards.sol [hunk 24]
```diff
@@ -3,234 +3,504 @@ pragma solidity ^0.4.24;
...
+
+        if(_claimableLink){
+            require(msg.value == EPHEMERAL_ADDRESS_FEE, "A claimable link was generated but not enough ephemeral ether was sent!");
+            _giftStatus = Statuses.Deposited;
+            _sentToAddress = this;
+            // need to store the address of the ephemeral account and the card that it owns for claimable link functionality
+            ephemeralWalletCards[to] = tokenIdPointer;
+            to.transfer(EPHEMERAL_ADDRESS_FEE);
+        }
+
```
Comment (andygray) [inline contracts/RadiCards.sol]: remove blank line

File: contracts/RadiCards.sol (hunk 19 above)
Comment (SoIidarity) [reply contracts/RadiCards.sol]: good catch. will lint the contracts now to ensure all the formatting is consistent

File: contracts/RadiCards.sol [hunk 25]
```diff
@@ -3,234 +3,504 @@ pragma solidity ^0.4.24;
...
+        // Card is minted to the _sentToAddress. This is either this radicards contract(if claimableLink==true)
+        // and the creator chose to use the escrow for a claimable link or to the recipient EOA directly
+        uint256 _tokenId = _mint(_sentToAddress, cards[_cardIndex].tokenURI);
+
+        cards[_cardIndex].minted++;
+
+        // transfer the DAI to the benefactor
+        if(_donationAmount > 0){
+            address _benefactorAddress = benefactors[_benefactorIndex].ethAddress;
+            require(daiContract.transferFrom(msg.sender, _benefactorAddress, _donationAmount),"Sending to charity failed");
```
[the remark this answers is not in the document]
Comment (SoIidarity) [reply contracts/RadiCards.sol]: sure. I need to make sure my nomenclature is consistent.

File: contracts/RadiCards.sol (hunk 8 above)
Comment (SoIidarity) [reply contracts/RadiCards.sol]: there is a need as this defines what the units are for the `donationAmount` and `giftAmount`. the same variable will either store Wei or Atto for Eth and Dai respectively. so yes, I agree that we should add this to the event

**Comment by SoIidarity** [comment]
Time to merge...lets see how it goes 🚀
